app/assets/javascripts/discourse/app/controllers/sidebar-section-form.js

```diff
@@ -325,23 +363,111 @@ export default Controller.extend(ModalFunctionality, {
+  },
+
   get header() {
     return this.model.id
       ? "sidebar.sections.custom.edit"
       : "sidebar.sections.custom.add";
   },
 
+  @bind
+  reorder(linkIdFrom, linkIdTo, above) {
+    if (linkIdFrom === linkIdTo) {
+      return;
+    }
+    let fromLink = this.model.links.find(
+      (link) => link.objectId === linkIdFrom
+    );
+    if (!fromLink) {
+      fromLink = this.model.secondaryLinks.find(
+        (link) => link.objectId === linkIdFrom
+      );
+    }
+    this.model.links.removeObject(fromLink);
+    this.model.secondaryLinks?.removeObject(fromLink);
+
+    let toPosition = this.model.links.findIndex(
+      (link) => link.objectId === linkIdTo
+    );
+    if (toPosition === -1) {
+      fromLink.segment = "secondary";
+      toPosition = this.model.secondaryLinks.findIndex(
+        (link) => link.objectId === linkIdTo
+      );
+      if (above) {
+        this.model.secondaryLinks.insertAt(toPosition, fromLink);
+      } else {
+        this.model.secondaryLinks.insertAt(toPosition + 1, fromLink);
+      }
+    } else {
+      fromLink.segment = "primary";
+      if (above) {
+        this.model.links.insertAt(toPosition, fromLink);
+      } else {
+        this.model.links.insertAt(toPosition + 1, fromLink);
+      }
+    }
+  },
+
+  get canDelete() {
+    return this.model.id && !this.model.sectionType;
+  },
+
+  @bind
+  deleteLink(link) {
+    if (link.id) {
+      link._destroy = "1";
+    } else {
+      this.model.links.removeObject(link);
+      this.model.secondaryLinks.removeObject(link);
+    }
+  },
+
   actions: {
     addLink() {
-      this.model.links.pushObject(new SectionLink({ router: this.router }));
+      this.nextObjectId = this.nextObjectId + 1;
+      this.model.links.pushObject(
+        new SectionLink({ router: this.router, objectId: this.nextObjectId })
+      );
     },
 
-    deleteLink(link) {
-      if (link.id) {
-        link._destroy = "1";
-      } else {
-        this.model.links.removeObject(link);
-      }
+    addSecondaryLink() {
+      this.nextObjectId = this.nextObjectId + 1;
+      this.model.secondaryLinks.pushObject(
+        new SectionLink({ router: this.router, objectId: this.nextObjectId })
+      );
+    },
+
+    resetToDefault() {
+      return this.dialog.yesNoConfirm({
+        message: I18n.t("sidebar.sections.custom.reset_confirm"),
+        didConfirm: () => {
+          return ajax(`/sidebar_sections/reset/${this.model.id}`, {
+            type: "PUT",
+          })
+            .then((data) => {
+              const newSidebarSections = this.currentUser.sidebar_sections.map(
+                (section) => {
+                  if (section.id === data["sidebar_section"].id) {
+                    return data["sidebar_section"];
+                  }
+                  return section;
+                }
+              );
+              this.currentUser.set("sidebar_sections", newSidebarSections);
```

Comment:
Is it possible to do this in a more efficient way where we don't have to re-render all the sections when we're only changing one?

Reply:
Because Community section is first, I can simplify code with unshift/shift. However, this may still cause full re-render:
```javascript
              this.currentUser.sidebar_sections.shiftObject();
              this.currentUser.sidebar_sections.unshiftObject(
                data["sidebar_section"]
              );
```

Reply:
Yea I was thinking about this and realised that updating a section is not something done very often so the re-rendering should be fine.